### ISSUE TITLE
Overlapping DragDropTarget 

### ISSUE THREAD

**Issue by sivu** [ocornut/imgui#6183]
ImGui V1.89WIP (docking)

We have element that is a drag&drop target, and it also has a child that is drag&drop target for the same type of data. This causes problems with AcceptDragDropPayload().

(example is from out-of-my-head)
```
ImGui::BeginGroup();
for (int i=0;list.size();++i)
{
    ImGui::PushID(i);
    ImGui::BeginGroup();

    //draw list element

   ImGui::EndGroup();
   ImGui::PopID();
   if (ImGui::BeginDragDropTarget())
   {
         if (const ImGuiPayload* pPayload = ImGui::AcceptDragDropPayload(type))
         {
            //modify list item
         }
         ImGui::EndDragDropTarget();
   }
}
ImGui::EndGroup();
if (ImGui::BeginDragDropTarget())
{
         if (const ImGuiPayload* pPayload = ImGui::AcceptDragDropPayload(type))
         {
            //append new item to list
         }
         ImGui::EndDragDropTarget();
}
```

In ImGui::AcceptDragDropPayload:
```
const bool was_accepted_previously = (g.DragDropAcceptIdPrev == g.DragDropTargetId);

...snip...

payload.Delivery = was_accepted_previously && !IsMouseDown(g.DragDropMouseButton); // For extern drag sources affecting os window focus, it's easier to just test !IsMouseDown() instead of IsMouseReleased() 
if (!payload.Delivery && !(flags & ImGuiDragDropFlags_AcceptBeforeDelivery))
        return NULL;

return &payload;
```
This sets the payload.Delivery to true when you drop the element. this is set even if it returns NULL.

So after the element is dropped into the first DragDropTarget the second one still runs this code and will set payload.Delivery back to false even though it returns null.

in ImGui::EndFrame() there is this:
```
// Drag and Drop: Elapse payload (if delivered, or if source stops being submitted)
  if (g.DragDropActive)
  {
    bool is_delivered = g.DragDropPayload.Delivery;
    bool is_elapsed = (g.DragDropPayload.DataFrameCount + 1 < g.FrameCount) && ((g.DragDropSourceFlags & ImGuiDragDropFlags_SourceAutoExpirePayload) || !IsMouseDown(g.DragDropMouseButton));
    if (is_delivered || is_elapsed)
      ClearDragDrop();
  }
```

is_delivered will be false because of the second call to AcceptDragDropPayload even though it didn’t accept it.

So far the solution for us has been to clear drag&drop payload if AcceptDragDropPayload returns the payload.

(disclaimer: acting as a proxy here for a co-worker, so if the report is invalid, it is his fault, not mine)

**Comment by ocornut** [comment]
Thank you for the detailed post.
Interestingly, the suggested solution to #5817 involved the same change that your suggested.

I'll build myself a repro based on this code to think about it.

**Comment by ocornut** [comment]
I have pushed two fixes in e9743d8
And tests https://github.com/ocornut/imgui_test_engine/commit/b5d1ecc8ac2ef4c1c24a8e74775e5e9cfdad8142

I didn't expect using same data type when adding the overlapping thing, so that should fix it.

This should cherry-pick in docking with one trivial conflict which is the mismatching version bump.

Note that I put the code in EndDragDropTarget() as it was simpler but it is possible that very custom drag and drop schemes (using internals) would be less affected if the clear was done in two steps, one in the successful AcceptDragDropPayload() (marking drag and drop as inactive) and then actual payload can be cleared in EndDragDropTarget() or EndFrame().

(I was also surprised seeing drag and drop function used on `BeginGroup()`/`EndGroup()` blocks. I'm not even sure it would work with `BeginDragDropSource()`. Groups are weird in the sense that their ID vary: they use the ID of active item is if it within the group. It's a precarious concept.)

**Comment by sivu** [comment]
Thank you for a quick reply! We will test the fix and report back.

Interesting that drag&drop was not meant to be used with BeginGroup/EndGroup.  I think i've always used it like that.  Our edit element usually consists of multiple edit widgets (for example input text + combo/button(s)) so I wrap them in a group so I can consider them as a single element on higher-level. It makes the drop target area bigger and gives nicer highlight.
